lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx

```diff
@@ -0,0 +1,472 @@
+  , HSIC_XY_ ()
+  , HSIC_XX_ ()
+  , HSIC_YY_ ()
+  , R2HSICIndices_ ()
+  , permutationSize_(ResourceMap::GetAsUnsignedInteger("HSICEstimatorImplementation-PermutationSize"))
+  , isAlreadyComputedIndices_(false)
+  , isAlreadyComputedPValuesPermutation_(false)
+{
+  if(covarianceList_.getSize() != (inputSample_.getDimension() + outputSample_.getDimension())) throw InvalidDimensionException(HERE) << "The number of covariance momdels is the dimension of the input +1";
+  if(outputSample_.getDimension() != 1) throw InvalidDimensionException(HERE) << "The dimension of the output is 1.";
+  if(inputSample_.getSize() != outputSample_.getSize()) throw InvalidDimensionException(HERE) << "Input and output samples must have the same size";
+
+  weightFunction_  = SymbolicFunction("x", "1");
+}
+
+/* Virtual constructor */
+HSICEstimatorImplementation * HSICEstimatorImplementation::clone() const
+{
+  return new HSICEstimatorImplementation(*this);
+}
+
+/* Compute the weight matrix from the weight function */
+SquareMatrix HSICEstimatorImplementation::computeWeightMatrix(const Sample & Y) const
+{
+  const Sample wY(weightFunction_(Y));
+  const Scalar meanWY(wY.computeMean()[0]);
+  SquareMatrix mat(n_);
+  mat.setDiagonal(wY.asPoint() / meanWY);
+  return mat;
+}
+
+/* Compute a HSIC index (one marginal) by using the underlying estimator (biased or not) */
+Scalar HSICEstimatorImplementation::computeHSICIndex( const Sample & inSample, const Sample & outSample, const CovarianceModel & inCovariance, const CovarianceModel & outCovariance, const SquareMatrix & weightMatrix) const
+{
+  return estimatorType_.computeHSICIndex(inSample, outSample, inCovariance, outCovariance, weightMatrix);
+}
+
+/* Compute HSIC and R2-HSIC indices */
+void HSICEstimatorImplementation::computeIndices() const
+{
+  /* Compute weights */
+  const SquareMatrix W = computeWeightMatrix(outputSample_);
+
+  /* Init */
+  HSIC_XX_ = Point(inputDimension_);
+  HSIC_XY_ = Point(inputDimension_);
+  HSIC_YY_ = Point(1);
+
+  /* Loop over marginals: HSIC indices */
+  for(UnsignedInteger dim = 0; dim < inputDimension_; ++dim)
+  {
+    const Sample xdim(inputSample_.getMarginal(dim));
+    HSIC_XY_[dim] = computeHSICIndex(xdim, outputSample_, covarianceList_[dim], covarianceList_[inputDimension_], W);
+    HSIC_XX_[dim] = computeHSICIndex(xdim, xdim, covarianceList_[dim], covarianceList_[dim], W);
+  }
+  HSIC_YY_[0] = computeHSICIndex(outputSample_, outputSample_, covarianceList_[inputDimension_], covarianceList_[inputDimension_], W);
+
+  /* Compute R2-HSIC */
+  R2HSICIndices_ = Point(inputDimension_);
+  for(UnsignedInteger dim = 0; dim < inputDimension_; ++dim)
+  {
+    R2HSICIndices_[dim] = HSIC_XY_[dim] / sqrt(HSIC_XX_[dim] * HSIC_YY_[0]);
+  }
+
+  isAlreadyComputedIndices_ = true ;
+}
+
+/* Set permutation size */
+void HSICEstimatorImplementation::setPermutationSize(const UnsignedInteger B)
+{
+  permutationSize_ = B;
+  PValuesPermutation_ = Point();
```

Comment:
```suggestion
  PValuesPermutation_ = Point();
  isAlreadyComputedPValuesPermutation_ = false;
```

This is required so that the `ComputePValuesPermutation` can be run again after the set method is called. Otherwise the user will get an empty Point when calling the `getPValuesPermutation` method